crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,155 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
         }
     }
 
+    fn resolve_fields(&self, path: Option<&Path>) -> Option<(Ty, Vec<StructField>)> {
+        let def = path
+            .and_then(|path| self.module.resolve_path(self.db, &path).take_types())
+            .map(|def_id| def_id.resolve(self.db));
+
+        let def = if let Some(def) = def {
+            def
+        } else {
+            return None;
+        };
+
+        match def {
+            Def::Struct(s) => {
+                let fields: Vec<_> = self
+                    .db
+                    .struct_data(s.def_id())
+                    .variant_data
+                    .fields()
+                    .iter()
+                    .cloned()
+                    .collect();
+                Some((type_for_struct(self.db, s), fields))
+            }
+            Def::EnumVariant(ev) => {
+                let fields: Vec<_> = ev.variant_data(self.db).fields().iter().cloned().collect();
+                Some((type_for_enum_variant(self.db, ev), fields))
+            }
+            _ => None,
+        }
+    }
+
+    fn infer_tuple_struct(&mut self, path: Option<&Path>, subpats: &[PatId]) -> Ty {
+        let (ty, fields) = if let Some(x) = self.resolve_fields(path) {
+            x
+        } else {
+            return Ty::Unknown;
+        };
+
+        if fields.len() != subpats.len() {
+            return Ty::Unknown;
+        }
+
+        for (&subpat, field) in subpats.iter().zip(fields.iter()) {
+            let sub_ty = self.make_ty(&field.type_ref);
+            self.infer_pat(subpat, &Expectation::has_type(sub_ty));
+        }
+
+        ty
+    }
+
+    fn infer_struct(&mut self, path: Option<&Path>, subpats: &[FieldPat]) -> Ty {
+        let (ty, fields) = if let Some(x) = self.resolve_fields(path) {
+            x
+        } else {
+            return Ty::Unknown;
+        };
+
+        for subpat in subpats {
+            let matching_field = fields.iter().find(|field| field.name == subpat.name);
+
+            if let Some(field) = matching_field {
+                let typeref = &field.type_ref;
+                let sub_ty = self.make_ty(typeref);
+                self.infer_pat(subpat.pat, &Expectation::has_type(sub_ty));
+            }
+        }
+
+        ty
+    }
+
+    // TODO: Expectation should probably contain a Cow pointer to Ty?
+    // so that we can make new expectations of subtypes cheaply
+    fn infer_pat(&mut self, pat: PatId, expected: &Expectation) -> Ty {
```

Comment:
not sure if expectation is actually right here -- for expressions we have coercions to deal with, so `Expectation` is supposed to (later) distinguish between "needs the exact type", "needs to be coercible" etc., and for patterns that doesn't apply

Reply:
Right - should we just take a `&Ty` instead?